server/remote_cache/chunked_manifest/chunked_manifest.go

```diff
@@ -0,0 +1,214 @@
+	acRNProto, err := acResourceName(blobDigest, instanceName, digestFunction)
+	if err != nil {
+		return nil, err
+	}
+
+	arBytes, err := cache.Get(ctx, acRNProto)
+	if err != nil {
+		if status.IsNotFoundError(err) {
+			return nil, status.NotFoundErrorf("chunked manifest not found for blob %s, %w", blobDigest.GetHash(), err)
+		}
+		return nil, status.InternalErrorf("retrieve chunked manifest from AC: %w", err)
+	}
+
+	manifest, err := decodeFromActionResultBytes(arBytes, blobDigest, instanceName, digestFunction)
+	if err != nil {
+		return nil, err
+	}
+
+	rns := make([]*rspb.ResourceName, 0, len(manifest.ChunkDigests))
+	for _, chunkDigest := range manifest.ChunkDigests {
+		chunkRN := digest.NewResourceName(chunkDigest, manifest.InstanceName, rspb.CacheType_CAS, manifest.DigestFunction)
+		rns = append(rns, chunkRN.ToProto())
+	}
+
+	// TODO: We could eventually return the missing chunks, and have the client
+	// check other places for those, before failing. This adds complexity, and
+	// will be avoided in the MVP.
+	missing, err := cache.FindMissing(ctx, rns)
+	if err != nil {
+		return nil, err
+	}
+	if len(missing) > 0 {
+		return nil, status.NotFoundErrorf("manifest found with missing chunks %+v", missing)
+	}
+	return manifest, nil
+}
+
+func decodeFromActionResultBytes(arBytes []byte, blobDigest *repb.Digest, instanceName string, digestFunction repb.DigestFunction_Value) (*ChunkedManifest, error) {
+	ar := &repb.ActionResult{}
+	if err := proto.Unmarshal(arBytes, ar); err != nil {
+		return nil, status.InternalErrorf("unmarshal chunked manifest from ActionResult: %w", err)
+	}
+
+	if len(ar.GetOutputFiles()) == 0 {
+		return nil, status.InvalidArgumentError("chunked manifest is empty")
+	}
+
+	chunkDigests := make([]*repb.Digest, 0, len(ar.GetOutputFiles()))
+	for _, outputFile := range ar.GetOutputFiles() {
+		chunkDigests = append(chunkDigests, outputFile.GetDigest())
+	}
+
+	return &ChunkedManifest{
+		BlobDigest:     blobDigest,
+		ChunkDigests:   chunkDigests,
+		InstanceName:   instanceName,
+		DigestFunction: digestFunction,
+	}, nil
+}
+
+func (cm *ChunkedManifest) verifyChunks(ctx context.Context, cache interfaces.Cache) error {
+	hasher, err := digest.HashForDigestType(cm.DigestFunction)
+	if err != nil {
+		return status.InvalidArgumentErrorf("invalid digest function: %s", err)
+	}
+
+	var totalSize int64
+	for i, chunkDigest := range cm.ChunkDigests {
+		chunkRN := digest.NewResourceName(chunkDigest, cm.InstanceName, rspb.CacheType_CAS, cm.DigestFunction)
+		if err := chunkRN.Validate(); err != nil {
+			return status.InvalidArgumentErrorf("invalid chunk digest at index %d for blob %s: %s", i, cm.BlobDigest.GetHash(), err)
+		}
+
+		reader, err := cache.Reader(ctx, chunkRN.ToProto(), 0, 0)
+		if err != nil {
+			if status.IsNotFoundError(err) {
+				return status.InvalidArgumentErrorf("invalid manifest: chunk %d not found in the CAS: %s", i, chunkDigest.GetHash())
+			}
```

Comment:
nit: I wonder if we can just do one quick round of cache.FindMissing before going into this loop? It would add some network overhead, but it would save us from having to download each chunk to the current machine just to have the final chunk missing.
FindMissing should refresh all chunks TTL underneath, so it's one way to tell the cache to keep those blobs around while we are downloading them here.